data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java

```diff
@@ -411,29 +418,51 @@ private <T> Record<Event> getRecord(ConsumerRecord<String, T> consumerRecord, in
         return new Record<Event>(event);
     }
 
-    private void processRecord(final AcknowledgementSet acknowledgementSet, final Record<Event> record) {
+    private <T> void processRecord(final AcknowledgementSet acknowledgementSet, final Record<Event> record) {
         // Always add record to acknowledgementSet before adding to
         // buffer because another thread may take and process
         // buffer contents before the event record is added
         // to acknowledgement set
         if (acknowledgementSet != null) {
             acknowledgementSet.add(record.getData());
         }
+        long numRetries = 0;
+        final int retrySleepTimeMs = 100;
+        // Donot pause until half the poll interval time has expired
+        final long maxRetries = topicConfig.getMaxPollInterval().toMillis() / (2 * retrySleepTimeMs);
         while (true) {
             try {
                 bufferAccumulator.add(record);
                 break;
             } catch (Exception e) {
+                if (!paused && numRetries++ > maxRetries) {
+                    paused = true;
+                    consumer.pause(consumer.assignment());
```

Comment:
This is very critical logic. We should have a unit test to verify that we call `pause()` on these conditions.

Reply:
Added new test case to test pause/resume